File: tests/python/unit/test_unit_model.py
from model import MusicModel
import pytest
import torch

def test_model_forward():
  model = MusicModel()

  batch_size = 16
  num_frames = 1290
  mfcc_size = 13
  x = torch.rand((batch_size, num_frames, mfcc_size))
  num_classes_expected = 10

  model.eval()
  with torch.no_grad():
    y = model(x)
  assert(y.numel() == batch_size * num_classes_expected), "Invalid size of the model output."

def test_model_save(tmp_path):
  file_path = tmp_path / "file.pt"
  model = MusicModel()
  model.save(file_path)
  assert(file_path.exists()), "Model is not saved."


@pytest.fixture(params=[True, False])
def exist(request):
  return request.param

def test_model_load(tmp_path, exist):
  file_path = tmp_path / "file.pt"
  model_save = MusicModel()
  model_load = MusicModel()

  if exist:
    model_save.save(file_path)
    model_load.load(file_path)
    for k, v in model_save.state_dict().items():
      assert(torch.equal(v, model_load.state_dict()[k])), "Model is not loaded properly."
  else:
    with pytest.raises(FileNotFoundError, match=f"MusicModel.load: File: {file_path} does not exist."):
      model_load.load(file_path)

File: tests/cpp/unit/testUnitModel.cpp
#include "model.hpp"
#include <gtest/gtest.h>
#include <memory>
#include <torch/autograd.h>

class TestModel : public ::testing::Test
{
  protected:
    MusicModel* model;

    void SetUp() override
    {
      impl = std::make_shared<MusicModelImpl>();
      model = new MusicModel(impl);
    }

    void TearDown() override
    {
      delete model;
      model = nullptr;
    }

  private:
    std::shared_ptr<MusicModelImpl> impl;
};

TEST_F(TestModel, forward)
{
  long numFrames = 1290;
  long mfccSize = 13;
  torch::Tensor x = torch::rand({1, numFrames, mfccSize});
  long numClassesExpected = 10;

  torch::Tensor y = model->forward(x);
  ASSERT_EQ(y.numel(), numClassesExpected) << "Invalid size of the model output.";
}

TEST_F(TestModel, train)
{
  model->train(true);
  ASSERT_TRUE(model->get()->is_training()) << "Invalid train mode.";

  model->train(false);
  ASSERT_FALSE(model->get()->is_training()) << "Invalid train mode.";
}

TEST_F(TestModel, eval)
{
  model->train(true);
  model->eval();
  ASSERT_FALSE(model->get()->is_training()) << "Invalid evaluation mode.";
}

TEST_F(TestModel, save)
{
  fs::path filePath = fs::temp_directory_path() / "file.pt";

  model->save(filePath);
  EXPECT_TRUE(fs::exists(filePath)) << "Model is not saved.";

  fs::remove(filePath);
}

class TestExistParam : public ::testing::TestWithParam<bool>
{
};

TEST_P(TestExistParam, load)
{
  bool exists = GetParam();
  fs::path filePath = fs::temp_directory_path() / "file.pt";
  MusicModel modelSave;
  MusicModel modelLoad;

  if (exists)
  {
    modelSave.save(filePath);
    modelLoad.load(filePath);
    for (const auto &item : modelSave.get()->named_parameters())
    {
      const auto& name = item.key();
      const auto& tensorSave = item.value();
      const auto& tensorLoad = modelLoad.get()->named_parameters()[name];
      ASSERT_TRUE(torch::equal(tensorSave, tensorLoad)) << "Model is not loaded properly.";
    }
  }
  else
    EXPECT_THROW(modelLoad.load(filePath), std::invalid_argument);
  fs::remove(filePath);
}

INSTANTIATE_TEST_SUITE_P(
  TestModelWithParams,
  TestExistParam,
  ::testing::Values(true, false),
  [](const testing::TestParamInfo<bool> &info)
  {
    return ::testing::PrintToString(info.param);
  }
);
